File: ruby/say/say.rb
# Influenced heavily by:
# https://github.com/xavier/exercism-assignments/blob/master/ruby/say/say.rb

class Say

  UNITS = {
    3 => "billion",
    2 => "million",
    1 => "thousand"
  }

  TENS = {
    90 => "ninety",
    80 => "eighty",
    70 => "seventy",
    60 => "sixty",
    50 => "fifty",
    40 => "forty",
    30 => "thirty",
    20 => "twenty"
  }

  TEENS = {
    19 => "nineteen",
    17 => "seventeen",
    16 => "sixteen",
    15 => "fifteen",
    14 => "fourteen",
    13 => "thirteen",
    12 => "twelve",
    11 => "eleven",
    10 => "ten"
  }

  DIGITS = {
    9 => "nine",
    8 => "eight",
    7 => "seven",
    6 => "six",
    5 => "five",
    4 => "four",
    3 => "three",
    2 => "two",
    1 => "one",
    0 => "zero"
  }

  ACCEPTED_RANGE = 0...1000000000000

  def initialize(number)
    raise ArgumentError unless ACCEPTED_RANGE.include?(number)
    @number = number
  end

  def in_english
    number_to_words(@number)
  end

  private

  def number_to_words(number)
    return DIGITS[number]  if number < 10
    return TEENS[number] if number < 20
    return number_to_words_100(number) if number < 100
    return number_to_words_1000(number) if number < 1000
    return general_case(number)
  end

  def number_to_words_100(number)
    tens_digit, ones_digit = number.divmod(10)
    tens = TENS[tens_digit * 10]
    ones = DIGITS[ones_digit]
    ones_digit == 0 ? "#{tens}" : "#{tens}-#{ones}"
  end

  def number_to_words_1000(number)
    hundreds_digit, leftover_digits = number.divmod(100)
    hundreds = DIGITS[hundreds_digit]
    leftovers = number_to_words_100(leftover_digits)
    leftover_digits == 0 ? "#{hundreds} hundred" : "#{hundreds} hundred #{leftovers}"
  end

  def chunkify(number)
    chunks = []
    while number > 0
      number, chunk = number.divmod(1000)
      chunks << chunk
    end
    chunks
  end

  def general_case(number)
    result = ""
    chunks = chunkify(number)
    chunks.each_with_index do |chunk, index|
      val = chunk_for(chunk)
      units = units_for(index)
      if val != ""
        result.prepend("#{val} #{units} ")
      end
    end
    result.strip
  end

  def chunk_for(number)
    return "" if number == 0
    return DIGITS[number]  if number < 10
    return TEENS[number] if number < 20
    return number_to_words_100(number) if number < 100
    return number_to_words_1000(number) if number < 1000
  end

  def units_for(number)
    return "" if number == 0
    return UNITS[number]
  end

end

File: python/say/say.py
class Say:

    NUM_TO_WORD = {
        1000000000: 'billion',
        1000000: 'million',
        1000: 'thousand',
        90: 'ninety',
        80: 'eighty',
        70: 'seventy',
        60: 'sixty',
        50: 'fifty',
        40: 'forty',
        30: 'thirty',
        20: 'twenty',
        19: 'nineteen',
        17: 'seventeen',
        16: 'sixteen',
        15: 'fifteen',
        14: 'fourteen',
        13: 'thirteen',
        12: 'twelve',
        11: 'eleven',
        10: 'ten',
        9: 'nine',
        8: 'eight',
        7: 'seven',
        6: 'six',
        5: 'five',
        4: 'four',
        3: 'three',
        2: 'two',
        1: 'one',
        0: ''
    }

    def __init__(self, num):
        self.num = int(num)
        self._words = self.get_words(self.num)

    def get_words(self, num):
        self.raise_if_invalid(num)

        if num == 0:
            return 'zero'
        else:
            return ' '.join([self.convert_chunk_to_word(chunk, index)
                             for index, chunk
                             in enumerate(self.get_chunks(num))]).rstrip()

    def get_chunks(self, num):
        reversed_string = str(num)[::-1]
        reversed_chunks = ([reversed_string[i:i + 3]
                            for i in range(0, len(reversed_string), 3)])[::-1]
        return [int(x[::-1]) for x in reversed_chunks]

    def convert_chunk_to_word(self, chunk, i):
        hundreds_digit, left_over = divmod(chunk, 100)
        hundreds = self.convert_num_to_word(hundreds_digit)

        tens_digit, ones_digit = divmod(left_over, 10)
        if 10 < left_over < 20:
            tens = self.convert_num_to_word(left_over)
            ones = None
        else:
            tens = self.convert_num_to_word(tens_digit * 10)
            ones = self.convert_num_to_word(ones_digit)

        word_chunk = self.format_chunk(hundreds, tens, ones)
        units = self.get_units(len(self.get_chunks(self.num)) - 1 - i)

        if not word_chunk:
            return ''
        return ' '.join((word_chunk, units))

    def format_chunk(self, hundreds, tens, ones):
        chunk = ''
        if hundreds:
            chunk += hundreds + ' hundred '
        if hundreds and tens:
            chunk += 'and '
        if tens:
            chunk += tens
        if tens and ones:
            chunk += '-'
        if ones:
            chunk += ones
        return chunk

    def raise_if_invalid(self, num):
        if num < 0 or num > 999999999999:
            raise AttributeError

    def get_units(self, d):
        return self.convert_num_to_word(1000 ** d) if 1000 ** d > 1 else ''

    def convert_num_to_word(self, num):
        return self.NUM_TO_WORD[num]

    def in_english(self):
        return self._words


def say(num):
    return Say(num).in_english()
